File: backend/app/utils/news_text.py
"""Clean RSS/HTML news summaries for display."""

from __future__ import annotations

import html
import re

_IMG_SRC = re.compile(
    r"""src=["'](https?://[^"']+\.(?:jpg|jpeg|png|webp|gif)(?:\?[^"']*)?)["']""",
    re.IGNORECASE,
)
_BARE_IMG_URL = re.compile(
    r"""https?://\S+\.(?:jpg|jpeg|png|webp|gif)(?:\?\S*)?""",
    re.IGNORECASE,
)
_TAG = re.compile(r"<[^>]+>")
_URL = re.compile(r"https?://\S+")


def clean_news_summary(text: str | None, *, max_len: int = 220) -> str:
    if not text:
        return ""
    raw = html.unescape(text.strip())
    raw = re.sub(r"<(script|style)[^>]*>.*?</\1>", " ", raw, flags=re.IGNORECASE | re.DOTALL)
    # Prefer img alt text when present
    raw = re.sub(
        r"""<img[^>]+alt=["']([^"']+)["'][^>]*>""",
        r" \1 ",
        raw,
        flags=re.IGNORECASE,
    )
    raw = _TAG.sub(" ", raw)
    raw = _URL.sub(" ", raw)
    raw = html.unescape(raw)
    raw = re.sub(r"\s+", " ", raw).strip()
    if len(raw) > max_len:
        raw = raw[: max_len - 1].rstrip() + "…"
    return raw


def extract_news_thumbnail(text: str | None) -> str | None:
    if not text:
        return None
    raw = html.unescape(text)
    m = _IMG_SRC.search(raw)
    if m:
        return m.group(1)
    m = _BARE_IMG_URL.search(raw)
    return m.group(0) if m else None

File: frontend/src/utils/newsText.ts
/** Client-side fallback when API returns legacy raw summaries. */

export function cleanNewsSummary(raw: string | null | undefined, maxLen = 220): string {
  if (!raw) return ''
  let text = raw
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&amp;/g, '&')
  text = text.replace(/<(script|style)[^>]*>[\s\S]*?<\/\1>/gi, ' ')
  text = text.replace(/<img[^>]+alt=["']([^"']+)["'][^>]*>/gi, ' $1 ')
  text = text.replace(/<[^>]+>/g, ' ')
  text = text.replace(/https?:\/\/\S+/g, ' ')
  text = text.replace(/\s+/g, ' ').trim()
  if (text.length > maxLen) text = `${text.slice(0, maxLen - 1)}…`
  return text
}

export function newsSummaryDisplay(
  summary: string | null | undefined,
  hasUrl: boolean,
): string {
  const clean = cleanNewsSummary(summary)
  if (clean) return clean
  return hasUrl ? '本文为图文资讯，点击标题查看详情与配图' : '暂无文字摘要'
}
